### DISCUSSION TITLE
How is the DRM device number determined at OS startup?

### DISCUSSION THREAD

**Discussion by hnakayam** [notro/gud#25] [Q&A]
I am currently trying to run GUD on the Raspberry Pi OS. I have ported some kernel modules from the buildroot version and have succeeded in running GUD on the Raspberry Pi 4. However, I noticed that the DRM device number that should be specified sometimes changes (not constant) in case of Raspberry Pi OS.

Case A: DRM_DEV=0

```
pi@raspberrypi:~$ ls /sys/class/drm
card0           card0-HDMI-A-2     card1       version
card0-HDMI-A-1  card0-Writeback-1  renderD128
pi@raspberrypi:~$ 
```

Case B: DRM_DEV=1
```
pi@raspberrypi:~$ ls /sys/class/drm
card0  card1-HDMI-A-1  card1-Writeback-1  version
card1  card1-HDMI-A-2  renderD128
pi@raspberrypi:~$ 
```

note: I have changed /boot/config.txt setting by following buildroot version
```
pi@raspberrypi:~$ tail -15 /boot/config.txt

# Enable audio (loads snd_bcm2835)
dtparam=audio=on

[pi4]
# Enable DRM VC4 V3D driver on top of the dispmanx display stack
#dtoverlay=vc4-fkms-v3d
#max_framebuffers=2
dtoverlay=dwc2,dr_mode=peripheral
dtoverlay=vc4-kms-v3d-pi4,noaudio
max_framebuffers=2

[all]
#dtoverlay=vc4-fkms-v3d
enable_uart=1
pi@raspberrypi:~$
```

So my question is, in what order are DRM devices recognized by the Linux / Raspberry Pi OS? And are there any way to get a DRM device number that will always valid for GUD use? 

Thanks,

**Comment by notro** [comment]
Here's the answer: https://github.com/raspberrypi/linux/issues/5403#issuecomment-1484915019
Here's one solution: https://forums.raspberrypi.com/viewtopic.php?p=2105798&sid=f4c882ea11ceee60d802fb29cd603035#p2105798

**Comment by hnakayam** [comment]
I'm trying to find out which DRM device is suitable for using the GUD on the Raspberry Pi 4. Note: This script needs to be run as root user (to access the debugfs).

```
# clear DRM_DEV
DRM_DEV=

for i in $(ls /sys/kernel/debug/dri); do 
    if grep "^vc4" /sys/kernel/debug/dri/$i/name; then DRM_DEV=$i; fi
done

if [ -z $DRM_DEV ]; then
    echo "VC4 DRM device not found." 
else
    echo "VC4 DRM device number is $DRM_DEV" 
fi
```

As you know, debugfs is not always available, though...
